[SILOptimizer][DebugInfo] Preliminary support for DIExpression in SROA and Mem2Reg

mshockwave opened the pull request:
SROA and Mem2Reg now can leverage DIExpression -- op_fragment, more
specifically -- to generate correct debug info for optimized SIL. Some
important highlights:
  - The new swift::salvageDebugInfo, similar to llvm::salvageDebugInfo,
     tries to restore / transfer debug info from a deleted instruction.
     Currently I only implemented this for store instruction whose
     destination is an alloc_stack value.
 - DCE now does not remove `debug_value` instruction whose associating
     with a function argument SSA value that is not used elsewhere. Since
     that SSA value will not disappear so we should keep the debug info.

**NOTE**: This PR is only for code review. It depends on and _includes_ #38734  #38735 . Please merge this PR instead of the previous two.

mshockwave committed:
[SIL][DebugInfo] Bring advanced debug info to alloc_stack

The `alloc_stack` instruction now can carry advanced debug info like
source-variable-specific SIL location, type, and SIL DIExpression.

mshockwave committed:
[DebugInfo][SIL] Introduce the 'implicit' attribute for debug variable

Debug variables that are marked 'implicit' on its `debug_value`
instruction mean that they were generated by compiler. Optimizers are
free to remove them (if it becomes a dead code, for instance) even in
-Onone. Since they are barely used by users and keeping them might lead
to incorrect IRGen results.

mshockwave commented:
@swift-ci please test

adrian-prantl commented on include/swift/SIL/SILCloner.h:
```diff
@@ -748,9 +748,13 @@ SILCloner<ImplClass>::visitAllocStackInst(AllocStackInst *Inst) {
+  auto *NewInst =
+      getBuilder().createAllocStack(Loc, getOpType(Inst->getElementType()),
+                                    VarInfo, Inst->hasDynamicLifetime());
+  // Remap the auxiliary debug scope, if any
```
Nit: missing `.` at the end of sentence.

adrian-prantl commented on include/swift/SIL/SILCloner.h:
```diff
@@ -748,9 +748,13 @@ SILCloner<ImplClass>::visitAllocStackInst(AllocStackInst *Inst) {
+                                    VarInfo, Inst->hasDynamicLifetime());
+  // Remap the auxiliary debug scope, if any
+  if (VarInfo && VarInfo->Scope)
+    NewInst->setDebugVarScope(getOpScope(VarInfo->Scope));
```
Should this be factored into a helper function that does this for all instructions that carry a SILDebugVariable?
I.e., `remapVarInfo(NewInst, VarInfo);`

adrian-prantl commented on include/swift/SILOptimizer/Utils/DebugOptUtils.h:
```diff
@@ -76,6 +81,7 @@ eraseFromParentWithDebugInsts(SILInstruction *inst,
   // Just matching what eraseFromParentWithDebugInsts is today.
   if (nextII == inst->getIterator())
     ++nextII;
+  swift::salvageDebugInfo(inst);
```
Does this need to be part of the SROA PR or should this be in its own patch?

mshockwave replied:
My original thought was that if salvageDebugInfo is put in a separate patch (or PR), there will be no use case or test there, which is a little weird. What do you think?

adrian-prantl replied:
Could we write a standalone test using sil-opt that runs exactly one pass that would delete an instruction? If the answer is no, then let's just keep it.

mshockwave replied:
Currently we're calling salvageDebugInfo inside InstructionDeleter, which will affect multiple Passes. We _can_ first create a PR that only calls salvageDebugInfo in a single Pass then revise it to use InstructionDeleter in later PR...but I feel like it will be a little over-engineering

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -191,9 +191,11 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
                               SILType SILTy);
 
-  void handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
+  /// Return false if we fail to create the right DW_OP_LLVM_fragment operand
```
`.`

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -191,9 +191,11 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+  bool handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
                             SmallVectorImpl<uint64_t> &Operands);
-  void buildDebugInfoExpression(const SILDebugVariable &VarInfo,
+  /// Return false if we fail to create the desired !DIExpression
```
`.`

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -191,9 +191,11 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
-  void handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
+  /// Return false if we fail to create the right DW_OP_LLVM_fragment operand
+  bool handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
                             SmallVectorImpl<uint64_t> &Operands);
```
ArrayRef<uint64_t> ?

adrian-prantl replied:
Sorry, I didn't realize that.

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2355,7 +2356,9 @@ void IRGenDebugInfoImpl::handleFragmentDIExpr(
-  assert(Offset && FieldTy && "Non-fixed type?");
+  // Doesn't support non-fixed type right now
+  if (!Offset || !FieldTy)
+    return false;
```
I'd be curious to see an example of the IR here. Even for a non-fixed size type, I would expect us to emit some sort of fixed-size container that optionally contains a pointer to a heap/stack-allocated memory, and I would expect SROA to only split apart that fixed-size container and not the variable-sized part. We need to fix this right now, but maybe keep this in mind.

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2366,22 +2369,26 @@ void IRGenDebugInfoImpl::handleFragmentDIExpr(
+        return false;
       break;
     default:
       llvm_unreachable("Unrecoginized operator");
```
not your code: typo

adrian-prantl commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -4752,8 +4752,7 @@ void IRGenSILFunction::visitDebugValueInst(DebugValueInst *i) {
         Decl, RealTy, getTypeInfo(SILVal->getType()));
-  } else if (i->getFunction()->isBare() && !SILTy.hasArchetype() &&
-             !VarInfo->Name.empty()) {
+  } else if (!SILTy.hasArchetype() && !VarInfo->Name.empty()) {
```
why the change with isBare?

mshockwave replied:
A function is marked as bare if it's parsed from SIL. But I also saw cases where the function is coming from Swift code but the instruction (i.e. `DebugValueInst *I`) doesn't have an associated VarDecl. For example, the debug_value instruction was generated during the salvageDebugInfo. In those cases, we will restore the `DbgTy` from `RealTy` and `SILTy`.

adrian-prantl replied:
Under what circumstances does salvageDebugInfo drop the VarDecl? Shouldn't it carry it forward from the salvaged debug_value?

mshockwave replied:
We're actually retrieving a VarDecl of a debug_value from its attached SILLocation (`getLoc().getAsASTNode<VarDecl>();`). And in the case of debug_value created by salvageDebugInfo, its SILLocation is "<compiler generated>", which I believe we can not get it as VarDecl.

adrian-prantl replied:
Thanks! We discussed this offline and the piece of info I was missing that we are storing the type info in the new auxiliary info record, and this change effectively allows us to access that.

adrian-prantl commented on lib/SIL/IR/SILBasicBlock.cpp:
```diff
@@ -292,8 +292,17 @@ transferNodesFromList(llvm::ilist_traits<SILBasicBlock> &SrcTraits,
+          if (VarInfo->Scope)
+            DVI.setDebugVarScope(
+                ScopeCloner.getOrCreateClonedScope(VarInfo->Scope));
+        }
```
LLVM style usually omits {} for single statements.

mshockwave committed:
[SILOptimizer][DebugInfo] Preliminary support for DIExpression in SROA and Mem2Reg

SROA and Mem2Reg now can leverage DIExpression -- op_fragment, more
specifically -- to generate correct debug info for optimized SIL. Some
important highlights:
 - The new swift::salvageDebugInfo, similar to llvm::salvageDebugInfo,
   tries to restore / transfer debug info from a deleted instruction.
   Currently I only implemented this for store instruction whose
   destination is an alloc_stack value.
 - Since we now have source-variable-specific SIL location inside a
   `debug_value` instruction (and its friends), this patch teaches
   SILCloner and SILInliner to remap the debug scope there in addition
   to debug scope of the instruction.
 - DCE now does not remove `debug_value` instruction whose associating
   with a function argument SSA value that is not used elsewhere. Since
   that SSA value will not disappear so we should keep the debug info.

mshockwave commented:
I just found that I might can split the changes regarding debug scope here to a separate PR. I'll update this accordingly later

mshockwave commented:
> I just found that I might can split the changes regarding debug scope here to a separate PR. I'll update this accordingly later

Since that part of changes don't really affect the feature here, I'll put it in another PR that is not depending on the current series of patches
@swift-ci please test

mshockwave commented:
@swift-ci please test

mshockwave commented:
> > I just found that I might can split the changes regarding debug scope here to a separate PR. I'll update this accordingly later
> 
> Since that part of changes don't really affect the feature here, I'll put it in another PR that is not depending on the current series of patches

I was wrong about that: Without the debug scope changes it will fail on a pretty common optimization pattern (specialization on implicit-generated enum functions) despite passing all tests, so I'm bringing them back and add corresponding test cases. I've also addressed all the feedbacks on that part before they were previously detached from this PR.
Sorry for this.

@swift-ci please test

adrian-prantl reviewed (approved):
Great, I think you've addressed all my concerns.

mshockwave commented:
So the failure on Linux buildbot was caused by uninitialized struct type information during IRGen. I think it only happens if a debug_value instruction is the first ever SIL instruction that references a certain struct type, in which case IRGen hasn't translated that type into its internal representation (e.g. RecordTypeInfo).

A more sophisticated solution will be asking IRGen to resolve the type when this happens. But since this scenario is pretty rare and I don't want to add more complexity into this PR, for now I simply bails out the (LLVM) debug intrinsic generation process if this occurs, similar to our current resolution on non-fixed type.

@swift-ci please test

adrian-prantl commented:
> A more sophisticated solution will be asking IRGen to resolve the type when this happens. But since this scenario is pretty rare and I don't want to add more complexity into this PR, for now I simply bails out the (LLVM) debug intrinsic generation process if this occurs, similar to our current resolution on non-fixed type.

Thanks for adding the comment explaining this to IRGenDebugInfo. LGTM!
